### PR TITLE
Mimimize Pandas usage and use Numpy types

### PR DESCRIPTION
**Pull request by jd**
carbonara: use numpy native types for computing
tests: fix gabbi test title
carbonara: simplify unserialize
carbonara: use sampling from SplitKey
storage: return SplitKey objects when listing split keys
utils: fix to_timestamps when value list is empty

### CODE REVIEW

**Comment by jd** [comment]
Pull-request updated, HEAD is now d83d4cdf8eea1b96d0ba4561e560afe35bb9d8dd

**Comment by jd** [comment]
Pull-request updated, HEAD is now 1dd53e251003d2766312e9f4cf42d06e2dccbd4c

**Comment by jd** [comment]
Pull-request updated, HEAD is now 0a22420228a8606922d9dc0dee0dc62fd7ee2dff

**Comment by jd** [comment]
Pull-request updated, HEAD is now d1f2e2cc928671061f2967b7a811b9f5d155ac04

**Comment by jd** [comment]
Pull-request updated, HEAD is now a17db4114c80333a86ca80d6cf52b689f3cd4f3e

File: gnocchi/rest/__init__.py [hunk 1]
```diff
@@ -145,7 +145,10 @@ def PositiveNotNullInt(value):
 
 
 def Timespan(value):
```
Comment (sileht) [inline gnocchi/rest/__init__.py]: Do we still need that?

File: gnocchi/storage/__init__.py [hunk 2]
```diff
@@ -254,10 +254,12 @@ def get_cross_metric_measures(metrics, from_timestamp=None,
         for metric in metrics:
             if aggregation not in metric.archive_policy.aggregation_methods:
                 raise AggregationDoesNotExist(metric, aggregation)
-            if (granularity is not None and granularity
-               not in set(d.granularity
-                          for d in metric.archive_policy.definition)):
-                raise GranularityDoesNotExist(metric, granularity)
+            if granularity is not None:
```
Comment (sileht) [inline gnocchi/storage/__init__.py]: unless change ?

File: gnocchi/archive_policy.py [hunk 3]
```diff
@@ -155,20 +158,25 @@ def __init__(self, granularity=None, points=None, timespan=None):
                 raise ValueError(
                     u"timespan ≠ granularity × points")
 
-        if granularity is not None and granularity <= 0:
-            raise ValueError("Granularity should be > 0")
+        if granularity is not None:
+            if not isinstance(granularity, numpy.timedelta64):
```
Comment (jd) [inline gnocchi/archive_policy.py]: Agreed. REST does the conversion but SQLAlchemy JSON type does not, it only unserialize the JSON and returns integer. So I can convert integer to Numpy type in SQLAlchemy but I'm not sure it's better than doing that here.

File: gnocchi/rest/__init__.py (hunk 1 above)
Comment (jd) [inline gnocchi/rest/__init__.py]: Yes in order to return a proper error from voluptuous and not have https://github.com/gnocchixyz/gnocchi/issues/195

File: gnocchi/storage/__init__.py (hunk 2 above)
Comment (jd) [inline gnocchi/storage/__init__.py]: IIRC it was not because set uses `__hash__` and not `__eq__` for comparison and the `__hash__` function for `SplitKey` was shit. But I fixed it so it might work correctly again?
Anyway my break-based-loop _can_ be faster than using a set that iterate over all archive policy so I ask for a laisser-passer on this one. :-P

File: gnocchi/rest/__init__.py (hunk 1 above)
Comment (sileht) [inline gnocchi/rest/__init__.py]: gotcha

**Comment by jd** [comment]
@sileht is there anything you still want me to change?

**Comment by sileht** [comment]
No it's ok

File: gnocchi/storage/_carbonara.py [hunk 4]
```diff
@@ -117,9 +118,15 @@ def _store_metric_measures(metric, timestamp_key, aggregation,
                                granularity, data, offset=None, version=3):
         raise NotImplementedError
 
-    @staticmethod
-    def _list_split_keys_for_metric(metric, aggregation, granularity,
+    def _list_split_keys_for_metric(self, metric, aggregation, granularity,
                                     version=3):
+        return set(map(
+            functools.partial(carbonara.SplitKey, sampling=granularity),
+            utils.to_timestamps(self._list_split_keys(
```
Comment (chungg) [inline gnocchi/storage/_carbonara.py]: this seems strange. _list_split_keys will give back epoch time in seconds, then to_timestamps will convert to nanoseconds... shouldn't we just pass _list_split_keys straight to SplitKey?

File: gnocchi/carbonara.py [hunk 5]
```diff
@@ -407,6 +410,10 @@ def __init__(self, value, sampling):
             self.key = value.key
         elif isinstance(value, pandas.Timestamp):
             self.key = value.value / 10e8
+        elif isinstance(value, numpy.datetime64):
```
Comment (chungg) [inline gnocchi/carbonara.py]: what path would give splitkey a datetime64? having a little trouble figuring this out.

File: gnocchi/storage/_carbonara.py (hunk 4 above)
Comment (jd) [inline gnocchi/storage/_carbonara.py]: `_list_split_keys` returns time in seconds with nanoseconds precision.

> shouldn't we just pass _list_split_keys straight to SplitKey?

SplitKey wants a timestamps (= e.g `numpy.dateteim64`) as first arg. I can use `numpy.datetime64(int(x * 10e8, 'ns')` but it's what `to_timestamps` will do so I thought it's be simpler to just use it.

Comment (chungg) [inline gnocchi/storage/_carbonara.py]: the following is my understanding.

`+    def _list_split_keys(self, metric, aggregation, granularity, version=3):
          key = self._metric_key(metric)
          if not self._client.exists(key):
              raise storage.MetricDoesNotExist(metric)
         split_keys = set()
         hashes = self._client.hscan_iter(
             key, match=self._aggregated_field_for_split(aggregation, '*',
                                                         granularity, version))
         for f, __ in hashes:
             meta = f.decode("utf8").split(self.FIELD_SEP, 1)
             split_keys.add(meta[0])
         return split_keys`

this seems to be just returning a set of strings representing epoch time in seconds. if we pass it string to splitkey, it will correctly convert it to epoch time in seconds cast to float: https://github.com/gnocchixyz/gnocchi/blob/master/gnocchi/carbonara.py#L411

it seems to_timestamps unnecessarily converts to nanosecond float, https://github.com/gnocchixyz/gnocchi/blob/master/gnocchi/utils.py#L129... maybe i'm missing something.

Comment (jd) [inline gnocchi/storage/_carbonara.py]: Yes, you miss that `gnocchi.carbonara.SplitKey` has been changed in this patch. :)

Comment (chungg) [inline gnocchi/storage/_carbonara.py]: which patch? my comments stop at 2nd commit https://github.com/gnocchixyz/gnocchi/pull/197/commits/0641fd109ac4875d281bfe72d48c672757815ff5

Comment (chungg) [inline gnocchi/storage/_carbonara.py]: i also noticed, the 2nd commit adds 

```
 +        elif isinstance(value, numpy.datetime64):
 +            self.key = (
 +                (value - UNIX_UNIVERSAL_START64) / numpy.timedelta64(1, 's')
 +            )
```

but it disappears eventually.

Comment (jd) [inline gnocchi/storage/_carbonara.py]: https://github.com/gnocchixyz/gnocchi/pull/197/commits/a130bb4ef88047e72750087633f16d79251f1417#diff-d99450bd1e2fa5016cec0a79f33d77afL408

**Comment by jd** [comment]
Pull-request updated, HEAD is now 97b144470ea290110102bba2a8b18b929c7b09dd

**Comment by jd** [comment]
Pull-request updated, HEAD is now 0af9451b2a0ee29f718c1f7e1750ab0161144fc6

**Comment by jd** [comment]
Pull-request updated, HEAD is now 2b95b6282ecce36f5208351055bf7a83c4fdc17c

File: gnocchi/carbonara.py [hunk 6]
```diff
@@ -408,22 +406,17 @@ class SplitKey(object):
...
-            self.key = value.value / 10e8
-        elif isinstance(value, numpy.datetime64):
-            self.key = (
-                (value - UNIX_UNIVERSAL_START64) / numpy.timedelta64(1, 's')
-            )
         else:
-            self.key = float(value)
+            self.key = value
 
         self.sampling = sampling
```
Comment (chungg) [inline gnocchi/carbonara.py]: is there a reason you are changing sampling to be a numpy.timedelta and not just float representing seconds? or i guess is there a reason why it only accepts timedelta as sampling? is it possible to split that into another patch to possible shrink this patch?

Comment (jd) [inline gnocchi/carbonara.py]: We definitely want a numpy type. But I'll try to split that out.

**Comment by jd** [comment]
I need to rework the whole branch anyway since the first patch has been heavily modified in https://github.com/gnocchixyz/gnocchi/pull/208

I'll let that branch be for the time being as a "backup" but it's definitely not mergeable now. :)
